### PR TITLE
Add tensor step and capturable support to rmsprop

### PR DESCRIPTION
**Pull request by mlazos**
Towards fixing https://github.com/pytorch/pytorch/issues/115679
Fixes RMSprop step update while compiling

Adds capturable support to RMSprop

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire

### CODE REVIEW

File: torch/optim/rmsprop.py [hunk 1]
```diff
@@ -193,7 +207,9 @@ def step(self, closure=None):
             the gradient is normalized by an estimation of its variance
         weight_decay (float, optional): weight decay (L2 penalty) (default: 0)
         {_foreach_doc}
+        {_capturable_doc}
         {_maximize_doc}
+        {_capturable_doc}
```
Comment (janeyx99) [inline torch/optim/rmsprop.py:212]: is the duplication an accident here?

File: torch/optim/rmsprop.py [hunk 2]
```diff
@@ -267,13 +293,22 @@ def _single_tensor_rmsprop(
...
     maximize: bool,
     differentiable: bool,
+    capturable: bool,
     has_complex: bool,
 ):
 
+    # If compiling, the compiler will handle cudagraph checks, see note [torch.compile x capturable]
+    if not torch._utils.is_compiling() and capturable:
+        assert all(p.is_cuda and step.is_cuda for p, step in zip(params, state_steps)), \
+            "If capturable=True, params and state_steps must be CUDA tensors."
```
Comment (janeyx99) [inline torch/optim/rmsprop.py:303]: or XLA tensors (see nadam.py)

File: torch/optim/rmsprop.py [hunk 3]
```diff
@@ -325,17 +361,24 @@ def _multi_tensor_rmsprop(
...
     if len(params) == 0:
         return
 
     assert not differentiable, "_foreach ops don't support autograd"
 
-    grouped_tensors = Optimizer._group_tensors_by_device_and_dtype([params, grads, square_avgs, grad_avgs, momentum_buffer_list])
+    # If compiling, the compiler will handle cudagraph checks, see note [torch.compile x capturable]
+    if not torch._utils.is_compiling() and capturable:
+        assert all(p.is_cuda and step.is_cuda for p, step in zip(params, state_steps)), \
+            "If capturable=True, params and state_steps must be CUDA tensors."
```
Comment (janeyx99) [inline torch/optim/rmsprop.py:375]: or XLA tensors, see nadam.py

**Review by janeyx99 — COMMENTED** [review]
Change looks good overall, left some comments.
1. Could you update the PR body to describe that you're also adding capturable support to RMSprop?
2. looks like there's one more failing inductor test, likely because -lr will induce a .item() call. This probably requires a similar branching to adadelta in the code for Tensor lr.

File: test/optim/test_optim.py [hunk 4]
```diff
@@ -746,7 +746,7 @@ def test_rmsprop(self):
         state = {}
         p = torch.rand(10, requires_grad=True, dtype=torch.float64)
         grad = torch.rand(10, requires_grad=True, dtype=torch.float64)
-        state["step"] = 0
+        state["step"] = torch.zeros((), dtype=torch.float64)
```
[the remark this answers is not in the document]
Comment (mlazos) [reply test/optim/test_optim.py:749]: Yeah, it needs to be a tensor because of our assertions

File: test/inductor/test_compiled_optimizers.py [hunk 5]
```diff
@@ -424,7 +425,7 @@ def check_cudagraphs_ran(self):
     test_adamax_recompile = make_recompile_test(Adamax, lr=0.01)
     test_nadam_recompile = make_recompile_test(NAdam, lr=0.01)
     test_rprop_recompile = make_recompile_test(Rprop, kernel_count=1, lr=0.01)
-    test_rmsprop_recompile = make_recompile_test(RMSprop, kernel_count=1, lr=0.01)
+    test_rmsprop_recompile = make_recompile_test(RMSprop, lr=0.01)
```
[the remark this answers is not in the document]
Comment (mlazos) [reply test/inductor/test_compiled_optimizers.py:428]: The default value is 2

File: torch/optim/rmsprop.py [hunk 6]
```diff
@@ -369,6 +421,14 @@ def _multi_tensor_rmsprop(
...
-            torch._foreach_add_(grouped_params, grouped_momentum_buffer_list, alpha=-lr)
+            if capturable:
+                momentum_lr = torch._foreach_mul(grouped_momentum_buffer_list, -lr)
+                torch._foreach_add_(grouped_params, momentum_lr)
+            else:
+                torch._foreach_add_(grouped_params, grouped_momentum_buffer_list, alpha=-lr)
         else:
-            torch._foreach_addcdiv_(grouped_params, grouped_grads, avg, value=-lr)
+            if capturable:
+                torch._foreach_div_(avg, -lr)
```
Comment (janeyx99) [inline torch/optim/rmsprop.py:431]:
```suggestion
                torch._foreach_mul_(grouped_grads, -lr)
```
why not mul? i think mul should be faster than div

File: torch/testing/_internal/common_optimizers.py [hunk 7]
```diff
@@ -1739,6 +1753,16 @@ def _get_optim_inputs_including_global_cliquey_kwargs(
                 "test_mixed_device_dtype",
                 active_if=TEST_WITH_TORCHDYNAMO,
             ),
+            DecorateInfo(
+                toleranceOverride(
+                    {  # This override is needed because we split addcdiv if we are running in capturable mode
+                        # this results in more casts in fp16 => more error over 7 iterations.
+                        torch.float32: tol(atol=4e-04, rtol=0.01),
```
Comment (janeyx99) [inline torch/testing/_internal/common_optimizers.py:1760]: this is pretty large...

File: torch/optim/rmsprop.py [hunk 8]
```diff
@@ -369,6 +421,14 @@ def _multi_tensor_rmsprop(
         if momentum > 0:
             torch._foreach_mul_(grouped_momentum_buffer_list, momentum)
             torch._foreach_addcdiv_(grouped_momentum_buffer_list, grouped_grads, avg)
-            torch._foreach_add_(grouped_params, grouped_momentum_buffer_list, alpha=-lr)
+            if capturable:
```
Comment (janeyx99) [inline torch/optim/rmsprop.py:424]:
```suggestion
            if capturable and isinstance(lr, Tensor):
```
and everywhere else too

File: torch/optim/rmsprop.py (hunk 6 above)
Comment (mlazos) [reply torch/optim/rmsprop.py:431]: We actually had larger error when I did that - I don't think the grads are properly copied in some configurations. This shouldn't affect performance anyway as these are memory-bound kernels.

File: torch/testing/_internal/common_optimizers.py (hunk 7 above)
Comment (mlazos) [reply torch/testing/_internal/common_optimizers.py:1760]: yeah, not sure if we have other options here, this was in eager and I didn't see any issue with the way I decomposed the addcdiv. This can happen with fp16, as there are fewer casts with addcdiv vs the decomposition I did above.

File: torch/optim/rmsprop.py (hunk 6 above)
Comment (janeyx99) [reply torch/optim/rmsprop.py:435]: What do you mean the grads are not properly copied?

File: torch/testing/_internal/common_optimizers.py (hunk 7 above)
Comment (janeyx99) [reply torch/testing/_internal/common_optimizers.py:1760]: ah did u confirm that if you remove the fp16 Tensors in the params list this wouldn't be necessary?

File: torch/optim/rmsprop.py (hunk 6 above)
Comment (mlazos) [reply torch/optim/rmsprop.py:435]: Basically I think there are cases above where we are modifying the actual grads (if we aren't maximize, no weight decay), but regardless, the error is higher, and there is no performance degradation from this, so I think the div is fine.

File: torch/testing/_internal/common_optimizers.py (hunk 7 above)
Comment (mlazos) [reply torch/testing/_internal/common_optimizers.py]: Actually I was able to remove this after all, I think I only needed it in the earlier versions for some reason.

File: torch/optim/rmsprop.py (hunk 6 above)
Comment (janeyx99) [reply torch/optim/rmsprop.py:435]: this is concerning cuz the grads should not get modified in the update. This does not have to do with your code so no action needed here but if you’re noticing bad stuff elsewhere pls flag

**Review by janeyx99 — APPROVED** [review]
